mono/metadata/w32file.c

```diff
@@ -341,20 +343,18 @@ ves_icall_System_IO_MonoIO_FindCloseFile (HANDLE hnd)
 	return mono_w32file_find_close (hnd);
 }
 
-MonoString *
-ves_icall_System_IO_MonoIO_GetCurrentDirectory (gint32 *io_error)
+MonoStringHandle
+ves_icall_System_IO_MonoIO_GetCurrentDirectory (gint32 *io_error, MonoError *error)
 {
-	MonoError error;
-	MonoString *result;
+	MonoStringHandle result;
 	gunichar2 *buf;
 	int len, res_len;
 
 	len = MAX_PATH + 1; /*FIXME this is too smal under most unix systems.*/
 	buf = g_new (gunichar2, len);
 	
-	error_init (&error);
 	*io_error=ERROR_SUCCESS;
-	result = NULL;
+	result = NULL_HANDLE_STRING;
```

Comment:
`result = MONO_HANDLE_NEW(MonoString, NULL)`.

Sorry this part of the API is terrible.  You're doing a `MONO_HANDLE_ASSIGN (result, ...)` later, so `result` needs to be an allocated handle.  Confusingly `NULL_HANDLE_STRING` is just a NULL cast to a handle, so assigning to it won't work.

This is one is a must-fix.